gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/service/GobblinServiceManagerTest.java

```diff
@@ -180,50 +199,137 @@ public void cleanUp() throws Exception {
+    this.gobblinServiceManager.getFlowCatalog().getSpecStore().addSpec(runOnceSpec);
+
+    List<Spec> specs = (List<Spec>) this.gobblinServiceManager.getFlowCatalog().getSpecs();
+
+    Assert.assertEquals(specs.size(), 2);
+    if (specs.get(0).getUri().equals(spec.getUri())) {
+      Assert.assertEquals(specs.get(1).getUri(), runOnceSpec.getUri());
+    } else if (specs.get(0).getUri().equals(runOnceSpec.getUri())) {
+      Assert.assertEquals(specs.get(1).getUri(), spec.getUri());
+    } else {
+      Assert.fail();
+    }
+
+    // restart the service
+    serviceReboot();
+
+    // runOnce job should get deleted from the spec store after running and uncompilable job should stay
+    AssertWithBackoff.create().maxSleepMs(200L).timeoutMs(20000L).backoffFactor(1)
+        .assertTrue(input -> this.gobblinServiceManager.getFlowCatalog().getSpecs().size() == 1,
+            "Waiting for the runOnce job to finish");
+
+    specs = (List<Spec>) this.gobblinServiceManager.getFlowCatalog().getSpecs();
+    Assert.assertEquals(specs.get(0).getUri(), spec.getUri());
+    Assert.assertFalse(flowConfig.getSchedule().isRunImmediately());
+
+    // clean it
+    this.gobblinServiceManager.getFlowCatalog().remove(spec.getUri());
+    specs = (List<Spec>) this.gobblinServiceManager.getFlowCatalog().getSpecs();
+    Assert.assertEquals(specs.size(), 0);
+  }
+
+  @Test (dependsOnMethods = "testRestart")
+  public void testUncompilableJob() throws Exception {
+    FlowId flowId = new FlowId().setFlowGroup(TEST_GROUP_NAME).setFlowName(MockedSpecCompiler.UNCOMPILABLE_FLOW);
+    URI uri = FlowConfigResourceLocalHandler.FlowUriUtils.createFlowSpecUri(flowId);
+    FlowConfig flowConfig = new FlowConfig().setId(flowId)
+        .setTemplateUris(TEST_TEMPLATE_URI).setProperties(new StringMap(flowProperties));
 
+    RestLiResponseException exception = null;
+    try {
+      this.flowConfigClient.createFlowConfig(flowConfig);
+    } catch (RestLiResponseException e) {
+      exception = e;
+    }
+    Assert.assertEquals(exception.getStatus(), HttpStatus.BAD_REQUEST_400);
+    // uncompilable job should not be persisted
+    Assert.assertEquals(this.gobblinServiceManager.getFlowCatalog().getSpecs().size(), 0);
+    Assert.assertFalse(this.gobblinServiceManager.getScheduler().getScheduledFlowSpecs().containsKey(uri.toString()));
+  }
+
+  @Test (dependsOnMethods = "testUncompilableJob")
+  public void testRunOnceJob() throws Exception {
+    FlowConfig flowConfig = new FlowConfig().setId(TEST_FLOW_ID)
+        .setTemplateUris(TEST_TEMPLATE_URI).setProperties(new StringMap(flowProperties));
+
+    this.flowConfigClient.createFlowConfig(flowConfig);
+
+    // runOnce job should not be persisted
+    AssertWithBackoff.create().maxSleepMs(200L).timeoutMs(2000L).backoffFactor(1)
+        .assertTrue(input -> this.gobblinServiceManager.getFlowCatalog().getSpecs().size() == 0,
+          "Flow that was created is not " + "reflecting in FlowCatalog");
```

Comment:
Can we have a more descriptive message? I do not understand "not reflecting in FlowCatalog" really means.